## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [8]:
# your code
arr = np.random.rand(10)
print(arr)
arr[3:8] = -arr[3:8]
arr

[0.82546159 0.35364532 0.23311651 0.78828552 0.46885525 0.97065537
 0.8628247  0.10511101 0.95832371 0.45192839]


array([ 0.82546159,  0.35364532,  0.23311651, -0.78828552, -0.46885525,
       -0.97065537, -0.8628247 , -0.10511101,  0.95832371,  0.45192839])

**2.** Заменить максимальный элемент случайного массива на 0

In [10]:
# your code
arr[arr.argmax()] = 0
arr

array([ 0.82546159,  0.35364532,  0.23311651, -0.78828552, -0.46885525,
       -0.97065537, -0.8628247 , -0.10511101,  0.        ,  0.45192839])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [50]:
# your code
A = np.arange(8).reshape(2,4)
np.dstack(np.meshgrid(A[0],A[1]))

array([[[0, 4],
        [1, 4],
        [2, 4],
        [3, 4]],

       [[0, 5],
        [1, 5],
        [2, 5],
        [3, 5]],

       [[0, 6],
        [1, 6],
        [2, 6],
        [3, 6]],

       [[0, 7],
        [1, 7],
        [2, 7],
        [3, 7]]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [19]:
# your code
A = np.arange(12).reshape(4,3)
A = np.concatenate((A,A))
B = np.arange(4).reshape(2,2)

In [20]:
A,B

(array([[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8],
        [ 9, 10, 11],
        [ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8],
        [ 9, 10, 11]]),
 array([[0, 1],
        [2, 3]]))

In [30]:
A[[np.sum([len(np.intersect1d(lineA, lineB)) for lineB in B])>0 for lineA in A]]

array([[0, 1, 2],
       [3, 4, 5],
       [0, 1, 2],
       [3, 4, 5]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [35]:
# your code
A = np.array([[1,1,1],[1,1,2],[2,2,2],[2,2,2],[3,3,3],[3,3,1],[4,4,5],[5,5,5],[5,5,5],[6,6,6]])
A[[np.any(line!=line[0]) for line in A]]

array([[1, 1, 2],
       [3, 3, 1],
       [4, 4, 5]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [40]:
# your code
print(A)
np.unique(A,axis=0)

[[1 1 1]
 [1 1 2]
 [2 2 2]
 [2 2 2]
 [3 3 3]
 [3 3 1]
 [4 4 5]
 [5 5 5]
 [5 5 5]
 [6 6 6]]


array([[1, 1, 1],
       [1, 1, 2],
       [2, 2, 2],
       [3, 3, 1],
       [3, 3, 3],
       [4, 4, 5],
       [5, 5, 5],
       [6, 6, 6]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [60]:
# your code
X = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
res = 1
for i in range(min(len(X),len(X[0]))):
    if X[i][i]:
        res *= X[i][i]
res

3

In [66]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
np.extract(np.diagonal(X)!=0,np.diagonal(X)).prod()

np.int64(3)

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [73]:
# your code
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
sorted(x) == sorted(y)

True

In [72]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
x.sort()==y.sort()

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [74]:
# your code
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
cur_max = min(x)
for i in range(len(x)-1):
    if x[i] == 0 and x[i+1]>cur_max:
        cur_max = x[i+1]
cur_max

5

In [85]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
x[1:][x[:-1]==0].max()

np.int64(5)

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [90]:
# your code
x = [2, 2, 2, 3, 3, 3, 5]
counts = dict()
for el in x:
    if el not in counts.keys():
        counts[el] = 1
    else:
        counts[el] += 1
counts.keys(), counts.values()

(dict_keys([2, 3, 5]), dict_values([3, 3, 1]))

In [87]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
np.unique(x, return_counts=True)

(array([2, 3, 5]), array([3, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [99]:
# your code
from scipy.spatial.distance import cdist
from time import time
A=[(1,2),(3,4),(0,0)]
B=[(1,1),(6,5),(32,1)]
start = time()
Y = cdist(A, B, 'euclidean')
print(time()-start)
Y

0.0


array([[ 1.        ,  5.83095189, 31.01612484],
       [ 3.60555128,  3.16227766, 29.15475947],
       [ 1.41421356,  7.81024968, 32.01562119]])

In [107]:
A=[(1,2),(3,4),(0,0)]
B=[(1,1),(6,5),(32,1)]
def dist(pts:list):
    point1, point2 = pts
    return sum([(point1[i]-point2[i])**2 for i in range(len(point1))])**(1/2)
start = time()
res = []
for a in A:
    res.append([])
    for b in B:
        res[-1].append(dist([a,b]))
print(time()-start)
res

0.0


[[1.0, 5.830951894845301, 31.016124838541646],
 [3.605551275463989, 3.1622776601683795, 29.154759474226502],
 [1.4142135623730951, 7.810249675906654, 32.01562118716424]]

In [173]:
A=np.array([(1,2),(3,4),(0,0)])
B=np.array([(1,1),(6,5),(32,1)])


In [186]:
start = time()
res = np.sqrt(np.sum((A-B[:, np.newaxis])**2,axis=2)).T
print(time()-start)
res

0.0


array([[ 1.        ,  5.83095189, 31.01612484],
       [ 3.60555128,  3.16227766, 29.15475947],
       [ 1.41421356,  7.81024968, 32.01562119]])

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [77]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [187]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [208]:
# your code
average_calories = np.average(calorie_stats)
average_calories

np.float64(106.88311688311688)

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [195]:
# your code
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [197]:
# your code
median_calories = np.median(calorie_stats)
median_calories

np.float64(110.0)

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [201]:
# your code
[(p,np.percentile(calorie_stats,p)) for p in range(0,10)]

[(0, np.float64(50.0)),
 (1, np.float64(50.0)),
 (2, np.float64(50.0)),
 (3, np.float64(55.599999999999994)),
 (4, np.float64(70.0)),
 (5, np.float64(70.0)),
 (6, np.float64(75.6)),
 (7, np.float64(83.2)),
 (8, np.float64(90.0)),
 (9, np.float64(90.0))]

In [202]:
nth_percentile=4

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [205]:
# your code
len(np.extract(calorie_stats>60,calorie_stats))/len(calorie_stats)

0.961038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [207]:
# your code
calorie_std = np.std(calorie_stats)
calorie_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Более половины конкурентной продукции по калориям превышают значение 110, а наши CrunchieMunchies вообще содержат гораздо меньше калорий, чем продукты конкурентов - 96% конкурентов уступают по этому показателю. В среднем CrunchieMunchies содержат на 40 калорий меньше, чем конкуренты.
Mycrunch следует использовать эти факты, чтобы показать, что их продукт полезней, чем у конкурентов. С такими фактами даже громкие слова не понадобятся!